test/integration/local/test_model_context.py

```diff
@@ -0,0 +1,118 @@
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import os
+import subprocess
+import sys
+import time
+
+import pytest
+import requests
+import torch
+from concurrent.futures import ThreadPoolExecutor
+import csv
+
+from integration import model_gpu_context_dir
+
+BASE_URL = "http://0.0.0.0:8080/"
+PING_URL = BASE_URL + "ping"
+INVOCATION_URL = BASE_URL + "models/model/invoke"
+GPU_COUNT = torch.cuda.device_count()
+GPU_IDS_EXPECTED = [i for i in range(GPU_COUNT)]
+
+
+@pytest.fixture(scope="module", autouse=True)
+def container(image_uri):
+    try:
+        if 'cpu' in image_uri:
+            pytest.skip("Skipping because tests running on CPU instance")
+
+        command = (
+            "docker run --gpus=all -p 8080:8080 "
+            "--name sagemaker-pytorch-inference-toolkit-context-test "
+            "-v {}:/opt/ml/model "
+            "{} serve"
+        ).format(model_gpu_context_dir, image_uri)
+
+        proc = subprocess.Popen(command.split(), stdout=sys.stdout, stderr=subprocess.STDOUT)
+
+        attempts = 0
+        while attempts < 10:
+            time.sleep(3)
+            try:
+                requests.get(PING_URL)
+                break
+            except Exception:
+                attempts += 1
+                pass
+        yield proc.pid
+
+    finally:
+        if 'cpu' in image_uri:
+            pytest.skip("Skipping because tests running on CPU instance")
+        subprocess.check_call("docker rm -f sagemaker-pytorch-inference-toolkit-context-test".split())
+
+
+def test_context_all_device_ids():
+    time.sleep(5)
+
+    gpu_ids_actual = []
+    with open(os.path.join(model_gpu_context_dir, 'code', 'model_fn_device_info.txt'), 'r') as f:
+        for line in f:
+            gpu_ids_actual.append(int(line))
+
+    gpu_ids_actual = list(set(gpu_ids_actual))
+    gpu_ids_actual.sort()
+
+    assert gpu_ids_actual == GPU_IDS_EXPECTED
+
+
+def test_same_pid_threadid():
+    time.sleep(5)
+    headers = {"Content-Type": "application/json"}
+    with ThreadPoolExecutor(max_workers=GPU_COUNT) as executor:
+        for i in range(GPU_COUNT):
+            executor.submit(send_request, b'input', headers)
+
+    time.sleep(10)
+    input_fn_device_info = read_csv("input_fn_device_info.csv")
+    output_fn_device_info = read_csv("output_fn_device_info.csv")
+    predict_fn_device_info = read_csv("predict_fn_device_info.csv")
+
+    assert len(input_fn_device_info) == len(output_fn_device_info) == len(predict_fn_device_info)
+
+    for input_fn_row, output_fn_row, predict_fn_row in zip(
+        input_fn_device_info, output_fn_device_info, predict_fn_device_info
+    ):
+
+        device_id_input_fn, pid_input_fn, threadid_input_fn = input_fn_row
```

Comment:
We cannot guarantee the sequence of `input_fn_device_info.csv`, `output_fn_device_info.csv` and `predict_fn_device_info.csv` are the same, right?

Eg:
input_fn_device_info.csv
```
pid=0 device_id=0
pid=1 device_id=1
```

output_fn_device_info.csv
```
pid=1 device_id=1
pid=0 device_id=0
```

in above, the behavior is correct but test will fail.

Reply:
Also, we want to check device_id consistency across all four functions, namely `model_fn`, `input_fn`, `output_fn` and `predict_fn`

Reply:
Acknowledged. Fixed in latest commit.